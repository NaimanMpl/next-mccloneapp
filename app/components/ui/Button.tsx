import React from 'react';

interface ButtonProps {
  label: string,
  backgroundColor: string,
  color: string,
  fullWidth?: boolean,
  loading?: boolean,
  submit?: boolean
}

const Button = ({ backgroundColor, color, loading = false, label, fullWidth = true, submit = true }: ButtonProps) => {
  const textColor = `text-${color}`;
  return (
    <button
      type={submit ? 'submit' : undefined}
      className={`mt-4 ${fullWidth ? 'w-full' : 'px-10'} ${backgroundColor} rounded-md py-2 font-semibold ${textColor}`}
      disabled={loading}
    >
      {label}
    </button>
  )
}

export default Button;