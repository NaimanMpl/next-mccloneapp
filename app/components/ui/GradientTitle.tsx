import { cn } from '@/lib/utils'
import { ReactNode } from 'react'

interface GradientTitleProps {
  children: ReactNode,
  className?: string
}

const GradientTitle = ({ children, className }: GradientTitleProps) => {
  return (
    <h2 
      className={cn(`pointer-events-none z-10 whitespace-pre-wrap bg-gradient-to-r from-[#000000] from-65% to-neutral-200/60 bg-clip-text text-center text-7xl font-semibold leading-none text-transparent dark:from-white dark:to-slate-900/10 mobile:text-5xl mobile:from-50%`, className)}
    >
      {children}
    </h2>
  )
}

export default GradientTitle