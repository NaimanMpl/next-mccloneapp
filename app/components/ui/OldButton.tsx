import React from 'react';

interface OldButtonProps {
  label: string,
  backgroundColor: string,
  color: string,
  fullWidth?: boolean,
  loading?: boolean,
  submit?: boolean
}

const OldButton = ({ backgroundColor, color, loading = false, label, fullWidth = true, submit = true }: OldButtonProps) => {
  const textColor = `text-${color}`;
  return (
    <button
      type={submit ? 'submit' : undefined}
      className={`${fullWidth ? 'w-full' : 'px-10'} ${backgroundColor} rounded-md py-2 font-semibold ${textColor}`}
      disabled={loading}
    >
      {label}
    </button>
  )
}

export default OldButton;