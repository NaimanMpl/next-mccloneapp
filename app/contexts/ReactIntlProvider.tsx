'use client';
import { ReactNode } from 'react';
import { IntlProvider } from 'react-intl';

const ReactIntlProvider = ({ children }: { children: ReactNode }) => {
  return <IntlProvider locale='fr'>{children}</IntlProvider>;
};

export default ReactIntlProvider;
