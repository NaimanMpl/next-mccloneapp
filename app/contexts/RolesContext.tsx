import {
  createContext,
  Dispatch,
  ReactNode,
  SetStateAction,
  useContext,
  useEffect,
  useState,
} from 'react';
import { RoleData } from '../models/role.model';
import { getRolesAndTheirUsers } from '../services/roleservice';

interface IRolesContext {
  roles: RoleData[];
  setRoles: Dispatch<SetStateAction<RoleData[]>>;
  loading: boolean;
}

const RolesContext = createContext<IRolesContext | undefined>(undefined);

export const RolesProvider = ({ children }: { children: ReactNode }) => {
  const [loading, setLoading] = useState(true);
  const [roles, setRoles] = useState<RoleData[]>([]);

  useEffect(() => {
    const fetchRoles = async () => {
      const rolesAndUsers = await getRolesAndTheirUsers();
      setRoles(rolesAndUsers);
      setLoading(false);
    };

    fetchRoles();
  }, []);

  return (
    <RolesContext.Provider value={{ roles, loading, setRoles }}>
      {children}
    </RolesContext.Provider>
  );
};

export const useRoles = () => {
  const context = useContext(RolesContext);
  if (context === undefined) {
    throw new Error('useRoles must be used within a RolesProvider');
  }
  return context;
};
