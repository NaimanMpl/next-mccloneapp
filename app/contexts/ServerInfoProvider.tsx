import {
  createContext,
  Dispatch,
  ReactNode,
  SetStateAction,
  useContext,
  useEffect,
  useState,
} from 'react';
import { Server } from '../models/server.model';
import { getServerInfo } from '../services/serverservice';

interface IServerInfoContext {
  loading: boolean;
  serverInfo: Server | undefined;
  setServerInfo: Dispatch<SetStateAction<Server | undefined>>;
}

const ServerInfoContext = createContext<IServerInfoContext | undefined>(
  undefined
);

export const ServerInfoProvider = ({ children }: { children: ReactNode }) => {
  const [serverInfo, setServerInfo] = useState<Server>();
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    const socket = new WebSocket(process.env.NEXT_PUBLIC_WEBSOCKET_ENDPOINT);
    socket.onmessage = (event) => {
      const data: Server = JSON.parse(event.data);
      setServerInfo(data);
    };
  }, []);

  useEffect(() => {
    const fetchServerInfo = async () => {
      const serverInfo = await getServerInfo();
      console.log(serverInfo);
      setServerInfo(serverInfo);
      setLoading(false);
    };

    setLoading(true);
    fetchServerInfo();
  }, []);

  return (
    <ServerInfoContext.Provider value={{ serverInfo, setServerInfo, loading }}>
      {children}
    </ServerInfoContext.Provider>
  );
};

export const useServerInfo = () => {
  const context = useContext(ServerInfoContext);
  if (!context) {
    throw new Error('useServerInfo must be used wihin a ServerInfoProvider');
  }
  return context;
};
