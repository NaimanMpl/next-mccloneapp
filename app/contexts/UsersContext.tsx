import { createContext, Dispatch, ReactNode, SetStateAction, useContext, useEffect, useState } from "react";
import { User } from "../models/user.model";
import { getUsers } from "../services/userservice";

interface IUsersContext {
  users: User[],
  loading: boolean,
  setUsers: Dispatch<SetStateAction<User[]>>
}

export const UsersContext = createContext<IUsersContext | undefined>(undefined);

export const UsersProvider = ({ children }: { children: ReactNode }) => {
  const [ users, setUsers ] = useState<User[]>([]);
  const [ loading, setLoading ] = useState(false);

  useEffect(() => {
    const fetchUsers = async () => {
      const users = await getUsers();
      setUsers(users);
      setLoading(false);
    }

    setLoading(true);
    fetchUsers();
  }, []);

  return (
    <UsersContext.Provider value={{ users, loading, setUsers }}>
      {children}
    </UsersContext.Provider>
  )
}

export const useUsers = () => {
  const context = useContext(UsersContext);
  if (context === undefined) {
    throw new Error("useUsers must be used within a UsersProvider");
  }
  return context;
}