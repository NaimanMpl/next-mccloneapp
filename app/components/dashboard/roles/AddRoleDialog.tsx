import { useAddRoleForm } from '@/app/hooks/useAddRoleForm';
import { Button } from '@/components/ui/button';
import { DialogContent, DialogDescription, DialogFooter, DialogHeader, DialogTitle } from '@/components/ui/dialog';
import { Form, FormField, FormItem } from '@/components/ui/form';
import { Input } from '@/components/ui/input';
import { Label } from '@/components/ui/label';

const AddRoleDialog = () => {

  const { FormSchema, form, onSubmit } = useAddRoleForm();

  return (
    <DialogContent>
      <DialogHeader>
        <DialogTitle>Ajouter un role</DialogTitle>
        <DialogDescription>Lorem ipsum dolor sit amet consectetur adipisicing elit. Vel, blanditiis!</DialogDescription>
      </DialogHeader>
      <Form {...form}>
        <form onSubmit={form.handleSubmit(onSubmit)}>
          <FormField
            control={form.control}
            name='name'
            render={({ field }) => (
              <FormItem>
                <Label>Nom du role</Label>
                <Input placeholder="ModÃ©rateur" {...field} />
              </FormItem>
            )}
          />
          <DialogFooter className='mt-4'>
            <Button type='submit'>Sauvegarder</Button>
          </DialogFooter>
        </form>
      </Form>
    </DialogContent>
  )
}

export default AddRoleDialog