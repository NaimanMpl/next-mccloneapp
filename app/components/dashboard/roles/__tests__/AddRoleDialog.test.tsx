import mockAxios from "@/__mocks__/axios";
import { RolesProvider } from "@/app/contexts/RolesContext";
import { Dialog } from "@/components/ui/dialog";
import { Toaster } from "@/components/ui/toaster";
import { render, screen, waitFor } from "@testing-library/react";
import user from '@testing-library/user-event';
import AddRoleDialog from "../AddRoleDialog";

jest.mock('axios');

describe('<AddRoleDialog />', () => {

  afterAll(() => {
    mockAxios.reset();
    user.setup();
  });

  it('should render successfully', () => {
    const dialog = render(
      <RolesProvider>
        <Dialog open>
          <AddRoleDialog />
        </Dialog>
      </RolesProvider>
    );

    expect(dialog).toMatchSnapshot();
  });

  it('should post a new role when submitting a valid form', async () => {

    const newRole = "Test Role";

    mockAxios.get.mockResolvedValueOnce([]);
    const submit = mockAxios.post.mockResolvedValueOnce(newRole);
    
    render(
      <RolesProvider>
        <Dialog open>
          <AddRoleDialog />
        </Dialog>
      </RolesProvider>
    );

    const [ roleNameInput ] = screen.getAllByRole('textbox');
    await user.type(roleNameInput, newRole);

    const submitButton = screen.getByText('Sauvegarder');
    await user.click(submitButton);

    await waitFor(() => {
      expect(submit).toHaveBeenCalledWith("/api/roles", { name: newRole });
    })
  });

  it('should throw an error toast if API call fails', async () => {
    const newRole = "Joueur";

    mockAxios.get.mockResolvedValueOnce([]);
    mockAxios.post.mockRejectedValueOnce(new Error("Fail to POST new role !"));
    
    render(
      <>
        <RolesProvider>
          <Dialog open>
            <AddRoleDialog />
          </Dialog>
        </RolesProvider>
        <Toaster />
      </>
    );

    const [ roleNameInput ] = screen.getAllByRole('textbox');
    await user.type(roleNameInput, newRole);

    const submitButton = screen.getByText('Sauvegarder');
    await user.click(submitButton);

    await waitFor(() => {
      expect(screen.getByText("Uh-oh")).toBeInTheDocument();
    });
  });

  it('should not submit the form if empty role name is given', async () => {
    
    mockAxios.get.mockResolvedValueOnce([]);
    mockAxios.post.mockRejectedValueOnce(new Error("Empty fields!"));

    render(
      <>
        <RolesProvider>
          <Dialog open>
            <AddRoleDialog />
          </Dialog>
        </RolesProvider>
        <Toaster />
      </>
    );

    const [ roleNameInput ] = screen.getAllByRole('textbox');
    await user.clear(roleNameInput);

    const submitButton = screen.getByText('Sauvegarder');
    await user.click(submitButton);

    await waitFor(() => {
      expect(roleNameInput).toHaveValue('');
      expect(screen.getByText('Pas si vite !')).toBeInTheDocument();
    });
    
  });

});