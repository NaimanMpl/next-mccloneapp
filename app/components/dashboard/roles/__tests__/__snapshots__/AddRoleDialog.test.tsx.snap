// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddRoleDialog /> should render successfully 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    data-scroll-locked="1"
    style="pointer-events: none;"
  >
    <span
      aria-hidden="true"
      data-aria-hidden="true"
      data-radix-focus-guard=""
      style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
      tabindex="0"
    />
    <div
      aria-hidden="true"
      data-aria-hidden="true"
    />
    <div
      aria-hidden="true"
      class="fixed inset-0 z-50 bg-black/80 data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0"
      data-aria-hidden="true"
      data-state="open"
      style="pointer-events: auto;"
    />
    <div
      aria-describedby="radix-:r2:"
      aria-labelledby="radix-:r1:"
      class="fixed left-[50%] top-[50%] z-50 grid w-full max-w-lg translate-x-[-50%] translate-y-[-50%] gap-4 border border-border bg-background p-6 shadow-lg duration-200 data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[state=closed]:slide-out-to-left-1/2 data-[state=closed]:slide-out-to-top-[48%] data-[state=open]:slide-in-from-left-1/2 data-[state=open]:slide-in-from-top-[48%] sm:rounded-lg"
      data-state="open"
      id="radix-:r0:"
      role="dialog"
      style="pointer-events: auto;"
      tabindex="-1"
    >
      <div
        class="flex flex-col space-y-1.5 text-center sm:text-left"
      >
        <h2
          class="text-lg font-semibold leading-none tracking-tight"
          id="radix-:r1:"
        >
          Ajouter un role
        </h2>
        <p
          class="text-sm text-muted-foreground"
          id="radix-:r2:"
        >
          Lorem ipsum dolor sit amet consectetur adipisicing elit. Vel, blanditiis!
        </p>
      </div>
      <form>
        <div
          class="space-y-2"
        >
          <label
            class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
          >
            Nom du role
          </label>
          <input
            class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
            name="name"
            placeholder="ModÃ©rateur"
            value=""
          />
        </div>
        <div
          class="flex flex-col-reverse sm:flex-row sm:justify-end sm:space-x-2 mt-4"
        >
          <button
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-semibold ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
            type="submit"
          >
            Sauvegarder
          </button>
        </div>
      </form>
      <button
        class="absolute right-4 top-4 rounded-sm opacity-70 ring-offset-background transition-opacity hover:opacity-100 focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 disabled:pointer-events-none data-[state=open]:bg-accent data-[state=open]:text-muted-foreground"
        type="button"
      >
        <svg
          class="lucide lucide-x h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 6 6 18"
          />
          <path
            d="m6 6 12 12"
          />
        </svg>
        <span
          class="sr-only"
        >
          Close
        </span>
      </button>
    </div>
    <span
      aria-hidden="true"
      data-aria-hidden="true"
      data-radix-focus-guard=""
      style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
      tabindex="0"
    />
  </body>,
  "container": <div
    aria-hidden="true"
    data-aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
